src/jit/compiler.cpp

```diff
@@ -516,63 +488,475 @@ var_types    Compiler::argOrReturnTypeForStruct(unsigned size, CORINFO_CLASS_HAN
     case 5:
     case 6:
     case 7:
-#endif // _TARGET_AMD64_
-    case 8:
-#else // !_TARGET_64BIT_
-    case 4:
-#endif // !_TARGET_64BIT_
-        info.compCompHnd->getClassGClayout(clsHnd, &gcPtr);
-        useType = getJitGCType(gcPtr);
+        useType = TYP_I_IMPL;
+        break;
+
+#endif // _TARGET_XARCH_
+#endif // _TARGET_64BIT_
+
+    case TARGET_POINTER_SIZE:
+        if (IsHfa(clsHnd))
+        {
+#ifdef _TARGET_64BIT_
+            var_types hfaType = GetHfaType(clsHnd);
+            // A structSize of 8 with IsHfa, we have two possiblities:
+            // An HFA of one double or an HFA of two floats
+            //
+            // Check and exclude the case of an HFA of two floats
+            if (hfaType == TYP_DOUBLE)
+            {
+                // We have an HFA of one double
+                useType = TYP_DOUBLE;
+            }  
+            else 
+            {
+                assert(hfaType == TYP_FLOAT);
+
+                // We have an HFA of two floats
+                // This should be passed or returned in two FP registers
+                useType = TYP_UNKNOWN;
+            }
+#else  // a 32BIT target
+            // A structSize of 4 with IsHfa, it must be an HFA of one float
+            useType = TYP_FLOAT;
+#endif
+        }
+        else
+        {
+            BYTE gcPtr = 0;
+            // Check if this pointer-sized struct is wrapping a GC object
+            info.compCompHnd->getClassGClayout(clsHnd, &gcPtr);
+            useType = getJitGCType(gcPtr);
+        }
         break;
 
     default:
-#if FEATURE_MULTIREG_RET
-        if (forReturn)
+        useType = TYP_UNKNOWN;
+        break;
+    }
+
+    return useType;
+}
+
+//-----------------------------------------------------------------------------
+// getArgTypeForStruct: 
+//     Get the type that is used to pass values of the given struct type.
+//     If you have already retrieved the struct size then it should be 
+//     passed as the optional third argument, as this allows us to avoid
+//     an extra call to getClassSize(clsHnd)
+//
+// Arguments:
+//    clsHnd       - the handle for the struct type
+//    wbPassStruct - An "out" argument with information about how 
+//                   the struct is to be passed
+//    structSize   - the size of the struct type, 
+//                   or zero if we should call getClassSize(clsHnd)
+//
+// Return Value:
+//    For wbPassStruct you can pass a 'nullptr' and nothing will be written
+//     or returned for that out parameter.
+//    When *wbPassStruct is SPK_PrimitiveType this method's return value
+//       is the primitive type used to pass the struct.
+//    When *wbPassStruct is SPK_ByReference this method's return value
+//       is TYP_UNKNOWN and the struct type is passed by reference to a copy
+//    When *wbPassStruct is SPK_ByValue or SPK_ByValueAsHfa this method's return value 
+//       can be TYP_STRUCT and the struct type is passed using multiple registers.
```

Comment:
Should this say "is passed using multiple registers, if sufficient argument registers are available, and is otherwise passed by value on the stack"?